wagtail/core/models/__init__.py

```diff
@@ -497,21 +507,27 @@ def save(self, clean=True, user=None, log_action=False, **kwargs):
             if not ('update_fields' in kwargs and 'slug' not in kwargs['update_fields']):
                 # see if the slug has changed from the record in the db, in which case we need to
                 # update url_path of self and all descendants
-                old_record = Page.objects.get(id=self.id)
+                old_record = Page.objects.get(id=self.id).specific
                 if old_record.slug != self.slug:
                     self.set_url_path(self.get_parent())
-                    update_descendant_url_paths = True
+                    slug_changed = True
                     old_url_path = old_record.url_path
                     new_url_path = self.url_path
 
         result = super().save(**kwargs)
 
-        if not is_new and update_descendant_url_paths:
+        if slug_changed:
             self._update_descendant_url_paths(old_url_path, new_url_path)
+            # Emit page_slug_changed signal on successful db commit
```

Comment:
Just wondering why this needs to be deferred to after the transaction?

This behaviour might not be obvious to a developer who is using this signal, so they might think that their logic is running in the transaction when it's not.

Reply:
Good question. I was being cautious here really... I see the sort of things you implement with signals as 'side effects' that aren't part of the same transaction, and an issue in a 'side effect' should prevent the core functionality from working. Issuing signals in a transaction also means that implementers are more restricted. For example, you can't reliably go and do things in a new thread until the database changes are committed.